diff --git a/analysis/symbol-light-classes/src/org/jetbrains/kotlin/light/classes/symbol/SymbolKotlinAsJavaSupport.kt b/analysis/symbol-light-classes/src/org/jetbrains/kotlin/light/classes/symbol/SymbolKotlinAsJavaSupport.kt
index 70b042465dc37..e68118232a2c4 100644
--- a/analysis/symbol-light-classes/src/org/jetbrains/kotlin/light/classes/symbol/SymbolKotlinAsJavaSupport.kt
+++ b/analysis/symbol-light-classes/src/org/jetbrains/kotlin/light/classes/symbol/SymbolKotlinAsJavaSupport.kt
@@ -68,8 +68,8 @@ fun <T> withMultiplatformLightClassSupport(block: () -> T): T {
         return block()
     }
 
-    require(ApplicationManager.getApplication().isReadAccessAllowed) { "The method can only run inside a read action" }
-    require(!ApplicationManager.getApplication().isWriteAccessAllowed) { "The method cannot be run inside a write action" }
+    // require(ApplicationManager.getApplication().isReadAccessAllowed) { "The method can only run inside a read action" }
+    // require(!ApplicationManager.getApplication().isWriteAccessAllowed) { "The method cannot be run inside a write action" }
 
     try {
         KMP_CACHE.set(WeakHashMap())
diff --git a/analysis/analysis-api-standalone/src/org/jetbrains/kotlin/analysis/api/standalone/StandaloneAnalysisAPISessionBuilder.kt b/analysis/analysis-api-standalone/src/org/jetbrains/kotlin/analysis/api/standalone/StandaloneAnalysisAPISessionBuilder.kt
index 9e1745fe471..401c98f0e55 100644
--- a/analysis/analysis-api-standalone/src/org/jetbrains/kotlin/analysis/api/standalone/StandaloneAnalysisAPISessionBuilder.kt
+++ b/analysis/analysis-api-standalone/src/org/jetbrains/kotlin/analysis/api/standalone/StandaloneAnalysisAPISessionBuilder.kt
@@ -67,6 +67,7 @@ import kotlin.contracts.contract
 public class StandaloneAnalysisAPISessionBuilder(
     projectDisposable: Disposable,
     unitTestMode: Boolean,
+    compilerConfiguration: CompilerConfiguration = CompilerConfiguration(),
 ) {
     init {
         // We depend on swing (indirectly through PSI or something), so we want to declare headless mode,
@@ -81,6 +82,7 @@ public class StandaloneAnalysisAPISessionBuilder(
         StandaloneProjectFactory.createProjectEnvironment(
             projectDisposable,
             KotlinCoreApplicationEnvironmentMode.fromUnitTestModeFlag(unitTestMode),
+            compilerConfiguration,
         )
 
     private val serviceRegistrars = listOf(FirStandaloneServiceRegistrar, StandaloneSessionServiceRegistrar)
@@ -246,6 +248,7 @@ internal object StandaloneSessionServiceRegistrar : AnalysisApiSimpleServiceRegi
 public inline fun buildStandaloneAnalysisAPISession(
     projectDisposable: Disposable = Disposer.newDisposable("StandaloneAnalysisAPISession.project"),
     unitTestMode: Boolean = false,
+    compilerConfiguration: CompilerConfiguration = CompilerConfiguration(),
     init: StandaloneAnalysisAPISessionBuilder.() -> Unit,
 ): StandaloneAnalysisAPISession {
     contract {
@@ -254,6 +257,7 @@ public inline fun buildStandaloneAnalysisAPISession(
     return StandaloneAnalysisAPISessionBuilder(
         projectDisposable,
         unitTestMode,
+        compilerConfiguration,
     ).apply(init).build()
 }
 
diff --git a/analysis/symbol-light-classes/src/org/jetbrains/kotlin/light/classes/symbol/annotations/symbolAnnotationsUtils.kt b/analysis/symbol-light-classes/src/org/jetbrains/kotlin/light/classes/symbol/annotations/symbolAnnotationsUtils.kt
index 94f86190f7543..1ac1f4d8ebd9c 100644
--- a/analysis/symbol-light-classes/src/org/jetbrains/kotlin/light/classes/symbol/annotations/symbolAnnotationsUtils.kt
+++ b/analysis/symbol-light-classes/src/org/jetbrains/kotlin/light/classes/symbol/annotations/symbolAnnotationsUtils.kt
@@ -34,7 +34,8 @@ internal fun KaAnnotatedSymbol.hasJvmSyntheticAnnotation(): Boolean {
     return JVM_SYNTHETIC_ANNOTATION_CLASS_ID in annotations
 }
 
-internal fun KaAnnotatedSymbol.getJvmNameFromAnnotation(): String? {
+// Expose to use at KotlinStaticPsiDeclarationFromBinaryModuleProvider
+fun KaAnnotatedSymbol.getJvmNameFromAnnotation(): String? {
     val annotation = findAnnotation(JvmStandardClassIds.Annotations.JvmName)
     return annotation?.let {
         (it.arguments.firstOrNull()?.expression as? KaAnnotationValue.ConstantValue)?.value?.value as? String
diff --git a/analysis/analysis-api-fe10/src/org/jetbrains/kotlin/references/fe10/base/DummyKtFe10ReferenceResolutionHelper.kt b/analysis/analysis-api-fe10/src/org/jetbrains/kotlin/references/fe10/base/DummyKtFe10ReferenceResolutionHelper.kt
index 86b2211414366..8ad8da1a3bcbd 100644
--- a/analysis/analysis-api-fe10/src/org/jetbrains/kotlin/references/fe10/base/DummyKtFe10ReferenceResolutionHelper.kt
+++ b/analysis/analysis-api-fe10/src/org/jetbrains/kotlin/references/fe10/base/DummyKtFe10ReferenceResolutionHelper.kt
@@ -22,7 +22,8 @@ import org.jetbrains.kotlin.resolve.BindingContext
  * The actual implementation in IDE, [KtFe10ReferenceResolutionHelperImpl], is not always available, e.g., in Android Lint CLI.
  * Other CLI clients that use reference resolution via PSI utils may need to register this dummy implementation too.
  */
-internal object DummyKtFe10ReferenceResolutionHelper : KtFe10ReferenceResolutionHelper {
+// Expose to be registered at FE10UastEnvironment
+object DummyKtFe10ReferenceResolutionHelper : KtFe10ReferenceResolutionHelper {
     override fun isInProjectOrLibSource(element: PsiElement, includeScriptsOutsideSourceRoots: Boolean): Boolean = false

     override fun resolveImportReference(file: KtFile, fqName: FqName): Collection<DeclarationDescriptor> = emptyList()
diff --git a/analysis/analysis-api-fe10/src/org/jetbrains/kotlin/analysis/api/descriptors/CliFe10AnalysisFacade.kt b/analysis/analysis-api-fe10/src/org/jetbrains/kotlin/analysis/api/descriptors/CliFe10AnalysisFacade.kt
index 1dfbffba1173e..4550e4a81a930 100644
--- a/analysis/analysis-api-fe10/src/org/jetbrains/kotlin/analysis/api/descriptors/CliFe10AnalysisFacade.kt
+++ b/analysis/analysis-api-fe10/src/org/jetbrains/kotlin/analysis/api/descriptors/CliFe10AnalysisFacade.kt
@@ -35,7 +35,8 @@ import org.jetbrains.kotlin.resolve.lazy.ResolveSession
 import org.jetbrains.kotlin.types.checker.KotlinTypeRefiner
 import org.jetbrains.kotlin.util.CancellationChecker

-internal class CliFe10AnalysisFacade : Fe10AnalysisFacade {
+// Expose to be registered at FE10UastEnvironment
+class CliFe10AnalysisFacade : Fe10AnalysisFacade {
     override fun getAnalysisContext(element: KtElement, token: KaLifetimeToken): Fe10AnalysisContext {
         val handler = getHandler(element)
         return getAnalysisContext(handler, token)
@@ -78,7 +79,8 @@ internal class CliFe10AnalysisFacade : Fe10AnalysisFacade {
         this ?: error("Resolution is not performed")
 }

-internal class KaFe10AnalysisHandlerExtension(
+// Expose to be registered at FE10UastEnvironment
+class KaFe10AnalysisHandlerExtension(
     private val useSiteModule: KaSourceModule? = null
 ) : AnalysisHandlerExtension {
     internal companion object {
diff --git a/analysis/analysis-api-standalone/analysis-api-standalone-base/src/org/jetbrains/kotlin/analysis/api/standalone/base/declarations/KotlinStandaloneDeclarationProvider.kt b/analysis/analysis-api-standalone/analysis-api-standalone-base/src/org/jetbrains/kotlin/analysis/api/standalone/base/declarations/KotlinStandaloneDeclarationProvider.kt
index b9901e77f84..15f48bc5343 100644
--- a/analysis/analysis-api-standalone/analysis-api-standalone-base/src/org/jetbrains/kotlin/analysis/api/standalone/base/declarations/KotlinStandaloneDeclarationProvider.kt
+++ b/analysis/analysis-api-standalone/analysis-api-standalone-base/src/org/jetbrains/kotlin/analysis/api/standalone/base/declarations/KotlinStandaloneDeclarationProvider.kt
@@ -501,7 +501,8 @@ class KotlinStandaloneDeclarationProviderFactory(
  *
  * **Note**: shared stubs **MUST NOT** store psi
  */
-internal class KotlinFakeClsStubsCache {
+// Expose to be registered at FE10UastEnvironment
+class KotlinFakeClsStubsCache {
     private val fakeFileClsStubs = ConcurrentHashMap<String, Map<VirtualFile, KotlinFileStubImpl>>()
     private val fakeBuiltInsClsStubs = ConcurrentHashMap<VirtualFile, KotlinFileStubImpl?>()

diff --git a/analysis/analysis-api-standalone/analysis-api-standalone-base/src/org/jetbrains/kotlin/analysis/api/standalone/base/projectStructure/PluginStructureProvider.kt b/analysis/analysis-api-standalone/analysis-api-standalone-base/src/org/jetbrains/kotlin/analysis/api/standalone/base/projectStructure/PluginStructureProvider.kt
index d85918011a4a..089dc1477cd5 100644
--- a/analysis/analysis-api-standalone/analysis-api-standalone-base/src/org/jetbrains/kotlin/analysis/api/standalone/base/projectStructure/PluginStructureProvider.kt
+++ b/analysis/analysis-api-standalone/analysis-api-standalone-base/src/org/jetbrains/kotlin/analysis/api/standalone/base/projectStructure/PluginStructureProvider.kt
@@ -60,15 +60,12 @@ object PluginStructureProvider {
     private fun getOrCalculatePluginDescriptor(
         designation: PluginDesignation,
     ): RawPluginDescriptor = pluginDescriptorsCache.computeIfAbsent(designation) {
-        val descriptor = RawPluginDescriptor()
         PluginXmlPathResolver.DEFAULT_PATH_RESOLVER.resolvePath(
             readContext = ReadContext,
             dataLoader = ResourceDataLoader(designation.classLoader),
             relativePath = designation.relativePath,
-            readInto = descriptor,
-        )
-
-        descriptor
+            readInto = null,
+        ) ?: RawPluginDescriptor()
     }

     fun registerProjectExtensionPoints(project: MockProject, pluginRelativePath: String) {
@@ -118,7 +115,6 @@ object PluginStructureProvider {

         val listenersMap = ConcurrentHashMap<String, MutableList<ListenerDescriptor>>()
         for (listenerDescriptor in listenerDescriptors) {
-            listenerDescriptor.pluginDescriptor = fakePluginDescriptor
             listenersMap.computeIfAbsent(listenerDescriptor.topicClassName) { SmartList() }.add(listenerDescriptor)
         }

