diff --git a/analysis/analysis-api-standalone/src/org/jetbrains/kotlin/analysis/api/standalone/StandaloneAnalysisAPISessionBuilder.kt b/analysis/analysis-api-standalone/src/org/jetbrains/kotlin/analysis/api/standalone/StandaloneAnalysisAPISessionBuilder.kt
index 38cbb45f8a77a..ff5446f89efd1 100644
--- a/analysis/analysis-api-standalone/src/org/jetbrains/kotlin/analysis/api/standalone/StandaloneAnalysisAPISessionBuilder.kt
+++ b/analysis/analysis-api-standalone/src/org/jetbrains/kotlin/analysis/api/standalone/StandaloneAnalysisAPISessionBuilder.kt
@@ -64,6 +64,7 @@ import kotlin.contracts.contract
 public class StandaloneAnalysisAPISessionBuilder(
     projectDisposable: Disposable,
     unitTestMode: Boolean,
+    compilerConfiguration: CompilerConfiguration = CompilerConfiguration(),
     classLoader: ClassLoader = MockProject::class.java.classLoader
 ) {
     init {
@@ -79,6 +80,7 @@ public class StandaloneAnalysisAPISessionBuilder(
         StandaloneProjectFactory.createProjectEnvironment(
             projectDisposable,
             KotlinCoreApplicationEnvironmentMode.fromUnitTestModeFlag(unitTestMode),
+            compilerConfiguration,
             classLoader = classLoader
         )
 
@@ -251,6 +253,7 @@ internal object StandaloneSessionServiceRegistrar : AnalysisApiSimpleServiceRegi
 public inline fun buildStandaloneAnalysisAPISession(
     projectDisposable: Disposable = Disposer.newDisposable("StandaloneAnalysisAPISession.project"),
     unitTestMode: Boolean = false,
+    compilerConfiguration: CompilerConfiguration = CompilerConfiguration(),
     classLoader: ClassLoader = MockProject::class.java.classLoader,
     init: StandaloneAnalysisAPISessionBuilder.() -> Unit,
 ): StandaloneAnalysisAPISession {
@@ -260,6 +263,7 @@ public inline fun buildStandaloneAnalysisAPISession(
     return StandaloneAnalysisAPISessionBuilder(
         projectDisposable,
         unitTestMode,
+        compilerConfiguration,
         classLoader
     ).apply(init).build()
 }
diff --git a/analysis/symbol-light-classes/src/org/jetbrains/kotlin/light/classes/symbol/annotations/symbolAnnotationsUtils.kt b/analysis/symbol-light-classes/src/org/jetbrains/kotlin/light/classes/symbol/annotations/symbolAnnotationsUtils.kt
index bd0d415acd96d..4aac8ae7831dd 100644
--- a/analysis/symbol-light-classes/src/org/jetbrains/kotlin/light/classes/symbol/annotations/symbolAnnotationsUtils.kt
+++ b/analysis/symbol-light-classes/src/org/jetbrains/kotlin/light/classes/symbol/annotations/symbolAnnotationsUtils.kt
@@ -32,7 +32,8 @@ import org.jetbrains.kotlin.resolve.deprecation.DeprecationLevelValue
 /**
  * @return [AnnotationUseSiteTargetFilter] which allows [this] and [NoAnnotationUseSiteTargetFilter] filter
  */
-internal fun AnnotationUseSiteTarget?.toOptionalFilter(): AnnotationUseSiteTargetFilter {
+// Expose to use at KotlinStaticPsiDeclarationFromBinaryModuleProvider
+fun AnnotationUseSiteTarget?.toOptionalFilter(): AnnotationUseSiteTargetFilter {
     if (this == null) return NoAnnotationUseSiteTargetFilter
 
     return annotationUseSiteTargetFilterOf(NoAnnotationUseSiteTargetFilter, toFilter())
@@ -51,7 +52,8 @@ internal fun KaAnnotatedSymbol.hasJvmSyntheticAnnotation(
     return hasAnnotation(JVM_SYNTHETIC_ANNOTATION_CLASS_ID, useSiteTargetFilter)
 }
 
-internal fun KaAnnotatedSymbol.getJvmNameFromAnnotation(
+// Expose to use at KotlinStaticPsiDeclarationFromBinaryModuleProvider
+fun KaAnnotatedSymbol.getJvmNameFromAnnotation(
     useSiteTargetFilter: AnnotationUseSiteTargetFilter = AnyAnnotationUseSiteTargetFilter,
 ): String? {
     val annotation = findAnnotation(JvmStandardClassIds.Annotations.JvmName, useSiteTargetFilter)
diff --git a/analysis/symbol-light-classes/src/org/jetbrains/kotlin/light/classes/symbol/annotations/AnnotationUseSiteTargetFilter.kt b/analysis/symbol-light-classes/src/org/jetbrains/kotlin/light/classes/symbol/annotations/AnnotationUseSiteTargetFilter.kt
index b82a339e7bfd5..34242d9ff6f96 100644
--- a/analysis/symbol-light-classes/src/org/jetbrains/kotlin/light/classes/symbol/annotations/AnnotationUseSiteTargetFilter.kt
+++ b/analysis/symbol-light-classes/src/org/jetbrains/kotlin/light/classes/symbol/annotations/AnnotationUseSiteTargetFilter.kt
@@ -7,11 +7,13 @@ package org.jetbrains.kotlin.light.classes.symbol.annotations

 import org.jetbrains.kotlin.descriptors.annotations.AnnotationUseSiteTarget

-internal fun interface AnnotationUseSiteTargetFilter {
+// Expose to use at KotlinStaticPsiDeclarationFromBinaryModuleProvider
+fun interface AnnotationUseSiteTargetFilter {
     fun isAllowed(useSiteTarget: AnnotationUseSiteTarget?): Boolean
 }

-internal fun AnnotationUseSiteTarget?.toFilter(): AnnotationUseSiteTargetFilter = when (this) {
+// Expose to use at KotlinStaticPsiDeclarationFromBinaryModuleProvider
+fun AnnotationUseSiteTarget?.toFilter(): AnnotationUseSiteTargetFilter = when (this) {
     null -> NoAnnotationUseSiteTargetFilter
     AnnotationUseSiteTarget.FIELD -> FieldAnnotationUseSiteTargetFilter
     AnnotationUseSiteTarget.FILE -> FileAnnotationUseSiteTargetFilter
@@ -68,4 +70,4 @@ internal object SetterParameterAnnotationUseSiteTargetFilter : AnnotationUseSite
 internal object PropertyDelegateFieldAnnotationUseSiteTargetFilter : AnnotationUseSiteTargetFilter {
     override fun isAllowed(useSiteTarget: AnnotationUseSiteTarget?): Boolean =
         useSiteTarget == AnnotationUseSiteTarget.PROPERTY_DELEGATE_FIELD
-}
\ No newline at end of file
+}
diff --git a/analysis/analysis-api-standalone/analysis-api-standalone-base/src/org/jetbrains/kotlin/analysis/api/standalone/base/projectStructure/PluginStructureProvider.kt b/analysis/analysis-api-standalone/analysis-api-standalone-base/src/org/jetbrains/kotlin/analysis/api/standalone/base/projectStructure/PluginStructureProvider.kt
index e5a743e6d03ef..0f3757f7b901e 100644
--- a/analysis/analysis-api-standalone/analysis-api-standalone-base/src/org/jetbrains/kotlin/analysis/api/standalone/base/projectStructure/PluginStructureProvider.kt
+++ b/analysis/analysis-api-standalone/analysis-api-standalone-base/src/org/jetbrains/kotlin/analysis/api/standalone/base/projectStructure/PluginStructureProvider.kt
@@ -49,6 +49,8 @@ object PluginStructureProvider {
     private class ResourceDataLoader(val classLoader: ClassLoader) : DataLoader {
         override val pool: ZipFilePool? get() = null
         override fun load(path: String): InputStream? = classLoader.getResource(path)?.openStream()
+        // KT-70264
+        fun load(path: String, pluginDescriptorSourceOnly: Boolean): InputStream? = load(path)
         override fun toString(): String = "resources data loader"
     }
 
