diff --git a/analysis/analysis-api-fir/src/org/jetbrains/kotlin/analysis/api/fir/components/KaFirExpressionTypeProvider.kt b/analysis/analysis-api-fir/src/org/jetbrains/kotlin/analysis/api/fir/components/KaFirExpressionTypeProvider.kt
index dc107d213ef0..d8db8458fd45 100644
--- a/analysis/analysis-api-fir/src/org/jetbrains/kotlin/analysis/api/fir/components/KaFirExpressionTypeProvider.kt
+++ b/analysis/analysis-api-fir/src/org/jetbrains/kotlin/analysis/api/fir/components/KaFirExpressionTypeProvider.kt
@@ -244,7 +244,7 @@ internal class KaFirExpressionTypeProvider(
                     is FirNamedArgumentExpression, is FirSpreadArgumentExpression ->
                         arg.psi == argumentExpression.parent
                     else ->
-                        arg.psi == argumentExpression
+                        arg.psi == argumentExpression.unwrap()
                 }
             }?.value ?: return null
         return if (firParameterForExpression.isVararg)
diff --git a/analysis/analysis-api-fe10/src/org/jetbrains/kotlin/analysis/api/descriptors/components/KaFe10SymbolRelationProvider.kt b/analysis/analysis-api-fe10/src/org/jetbrains/kotlin/analysis/api/descriptors/components/KaFe10SymbolRelationProvider.kt
index 158b7d8aafee..7b3299463dcd 100644
--- a/analysis/analysis-api-fe10/src/org/jetbrains/kotlin/analysis/api/descriptors/components/KaFe10SymbolRelationProvider.kt
+++ b/analysis/analysis-api-fe10/src/org/jetbrains/kotlin/analysis/api/descriptors/components/KaFe10SymbolRelationProvider.kt
@@ -6,8 +6,10 @@
 package org.jetbrains.kotlin.analysis.api.descriptors.components
 
 import com.intellij.openapi.project.Project
+import com.intellij.openapi.vfs.VirtualFile
 import com.intellij.psi.search.GlobalSearchScope
 import com.intellij.psi.search.ProjectScope
+import org.jetbrains.kotlin.analysis.api.KaExperimentalApi
 import org.jetbrains.kotlin.analysis.api.KaSession
 import org.jetbrains.kotlin.analysis.api.components.KaSymbolRelationProvider
 import org.jetbrains.kotlin.analysis.api.descriptors.KaFe10Session
@@ -167,6 +169,9 @@ internal class KaFe10SymbolRelationProvider(
             override val librarySources: KaLibrarySourceModule? = null
             override val isSdk: Boolean = false
             override val binaryRoots: Collection<Path> = listOf(libraryPath)
+
+            @KaExperimentalApi
+            override val binaryVirtualFiles: Collection<VirtualFile> = emptyList()
             override val directRegularDependencies: List<KaModule> = emptyList()
             override val directDependsOnDependencies: List<KaModule> = emptyList()
             override val transitiveDependsOnDependencies: List<KaModule> = emptyList()
diff --git a/analysis/analysis-api-impl-base/tests/org/jetbrains/kotlin/analysis/api/impl/base/test/configurators/AnalysisApiBaseTestServiceRegistrar.kt b/analysis/analysis-api-impl-base/tests/org/jetbrains/kotlin/analysis/api/impl/base/test/configurators/AnalysisApiBaseTestServiceRegistrar.kt
index a6479df9402b..26f1632fb34a 100644
--- a/analysis/analysis-api-impl-base/tests/org/jetbrains/kotlin/analysis/api/impl/base/test/configurators/AnalysisApiBaseTestServiceRegistrar.kt
+++ b/analysis/analysis-api-impl-base/tests/org/jetbrains/kotlin/analysis/api/impl/base/test/configurators/AnalysisApiBaseTestServiceRegistrar.kt
@@ -124,11 +124,15 @@ object AnalysisApiBaseTestServiceRegistrar : AnalysisApiTestServiceRegistrar() {
             testServices.environmentManager.getProjectEnvironment(),
         ).distinct()
 
+        val mainBinaryVirtualFiles = mainBinaryModules.flatMap { it.binaryVirtualFiles }.distinct()
+
         val sharedBinaryRoots = StandaloneProjectFactory.getVirtualFilesForLibraryRoots(
             sharedBinaryDependencies.flatMap { binary -> binary.binaryRoots },
             testServices.environmentManager.getProjectEnvironment()
         ).distinct()
 
+        val sharedBinaryVirtualFiles = sharedBinaryDependencies.flatMap { it.binaryVirtualFiles }.distinct()
+
         project.apply {
             registerService(KotlinAnnotationsResolverFactory::class.java, KotlinStandaloneAnnotationsResolverFactory(project, testKtFiles))
 
@@ -143,8 +147,8 @@ object AnalysisApiBaseTestServiceRegistrar : AnalysisApiTestServiceRegistrar() {
                 val declarationProviderFactory = KotlinStandaloneDeclarationProviderFactory(
                     project,
                     testKtFiles,
-                    binaryRoots = mainBinaryRoots,
-                    sharedBinaryRoots = sharedBinaryRoots,
+                    binaryRoots = mainBinaryRoots + mainBinaryVirtualFiles,
+                    sharedBinaryRoots = sharedBinaryRoots + sharedBinaryVirtualFiles,
                     skipBuiltins = testServices.moduleStructure.allDirectives.contains(NO_RUNTIME),
                     shouldBuildStubsForBinaryLibraries = shouldBuildStubsForBinaryLibraries,
                 )
diff --git a/analysis/analysis-api-standalone/analysis-api-standalone-base/src/org/jetbrains/kotlin/analysis/api/standalone/base/projectStructure/StandaloneProjectFactory.kt b/analysis/analysis-api-standalone/analysis-api-standalone-base/src/org/jetbrains/kotlin/analysis/api/standalone/base/projectStructure/StandaloneProjectFactory.kt
index bf24e33cfc8ba..e2ae99ecaeda9 100644
--- a/analysis/analysis-api-standalone/analysis-api-standalone-base/src/org/jetbrains/kotlin/analysis/api/standalone/base/projectStructure/StandaloneProjectFactory.kt
+++ b/analysis/analysis-api-standalone/analysis-api-standalone-base/src/org/jetbrains/kotlin/analysis/api/standalone/base/projectStructure/StandaloneProjectFactory.kt
@@ -31,6 +31,8 @@ import com.intellij.psi.search.ProjectScope
 import com.intellij.util.io.URLUtil.JAR_PROTOCOL
 import com.intellij.util.io.URLUtil.JAR_SEPARATOR
 import com.intellij.util.messages.ListenerDescriptor
+import org.jetbrains.kotlin.analysis.api.KaExperimentalApi
+import org.jetbrains.kotlin.analysis.api.KaImplementationDetail
 import org.jetbrains.kotlin.analysis.api.impl.base.util.LibraryUtils
 import org.jetbrains.kotlin.analysis.api.platform.projectStructure.KotlinModuleDependentsProvider
 import org.jetbrains.kotlin.analysis.api.platform.projectStructure.KotlinProjectStructureProvider
@@ -340,10 +342,13 @@ object StandaloneProjectFactory {
         .flatMap { it.getJavaRoots(environment) }
 
     fun createSearchScopeByLibraryRoots(
-        roots: Collection<Path>,
+        binaryRoots: Collection<Path>,
+        binaryVirtualFiles: Collection<VirtualFile>,
         environment: KotlinCoreProjectEnvironment,
     ): GlobalSearchScope {
-        val virtualFileUrls = getVirtualFileUrlsForLibraryRootsRecursively(roots, environment)
+        val virtualFileUrlsFromBinaryRoots = getVirtualFileUrlsForLibraryRootsRecursively(binaryRoots, environment)
+        val virtualFileUrlsFromBinaryVirtualFiles = getVirtualFileUrlsForLibraryRootsRecursively(binaryVirtualFiles)
+        val virtualFileUrls = virtualFileUrlsFromBinaryRoots + virtualFileUrlsFromBinaryVirtualFiles
 
         return object : GlobalSearchScope(environment.project) {
             override fun contains(file: VirtualFile): Boolean = file.url in virtualFileUrls
@@ -367,6 +372,17 @@ object StandaloneProjectFactory {
             }
         }
 
+    @OptIn(KaImplementationDetail::class)
+    private fun getVirtualFileUrlsForLibraryRootsRecursively(
+        binaryVirtualFiles: Collection<VirtualFile>
+    ): Set<String> =
+        buildSet {
+            for (vf in binaryVirtualFiles) {
+                LibraryUtils.getAllVirtualFilesFromRoot(vf, includeRoot = true)
+                    .mapTo(this) { it.url }
+            }
+        }
+
     fun getVirtualFilesForLibraryRoots(
         roots: Collection<Path>,
         environment: KotlinCoreProjectEnvironment,
@@ -417,10 +433,12 @@ object StandaloneProjectFactory {
         }
     }
 
+    @OptIn(KaExperimentalApi::class)
     private fun KaLibraryModule.getJavaRoots(
         environment: KotlinCoreProjectEnvironment,
     ): List<JavaRoot> {
-        return getVirtualFilesForLibraryRoots(binaryRoots, environment).map { root ->
+        val binaryRootsAsVirtualFiles = getVirtualFilesForLibraryRoots(binaryRoots, environment) + binaryVirtualFiles
+        return binaryRootsAsVirtualFiles.map { root ->
             JavaRoot(root, JavaRoot.RootType.BINARY)
         }
     }
diff --git a/analysis/analysis-api-standalone/src/org/jetbrains/kotlin/analysis/project/structure/builder/KaBinaryModuleBuilder.kt b/analysis/analysis-api-standalone/src/org/jetbrains/kotlin/analysis/project/structure/builder/KaBinaryModuleBuilder.kt
index d53fa589605e..b034ff7f67f4 100644
--- a/analysis/analysis-api-standalone/src/org/jetbrains/kotlin/analysis/project/structure/builder/KaBinaryModuleBuilder.kt
+++ b/analysis/analysis-api-standalone/src/org/jetbrains/kotlin/analysis/project/structure/builder/KaBinaryModuleBuilder.kt
@@ -5,11 +5,13 @@
 
 package org.jetbrains.kotlin.analysis.project.structure.builder
 
+import com.intellij.openapi.vfs.VirtualFile
 import java.nio.file.Path
 
 @KtModuleBuilderDsl
 public abstract class KtBinaryModuleBuilder : KtModuleBuilder() {
     private val binaryRoots: MutableList<Path> = mutableListOf()
+    private val binaryVirtualFiles: MutableList<VirtualFile> = mutableListOf()
 
     /**
      * Adds a [root] to the current library.
@@ -36,4 +38,14 @@ public abstract class KtBinaryModuleBuilder : KtModuleBuilder() {
     }
 
     protected fun getBinaryRoots(): List<Path> = binaryRoots.distinct()
+
+    public fun addBinaryVirtualFile(virtualFile: VirtualFile) {
+        binaryVirtualFiles.add(virtualFile)
+    }
+
+    public fun addBinaryVirtualFiles(virtualFiles: Collection<VirtualFile>) {
+        binaryVirtualFiles.addAll(virtualFiles)
+    }
+
+    protected fun getBinaryVirtualFiles(): List<VirtualFile> = binaryVirtualFiles.distinct()
 }
\ No newline at end of file
diff --git a/analysis/analysis-api-standalone/src/org/jetbrains/kotlin/analysis/project/structure/builder/KaLibraryModuleBuilder.kt b/analysis/analysis-api-standalone/src/org/jetbrains/kotlin/analysis/project/structure/builder/KaLibraryModuleBuilder.kt
index 3128be3f0c5f..dc8ab7624bed 100644
--- a/analysis/analysis-api-standalone/src/org/jetbrains/kotlin/analysis/project/structure/builder/KaLibraryModuleBuilder.kt
+++ b/analysis/analysis-api-standalone/src/org/jetbrains/kotlin/analysis/project/structure/builder/KaLibraryModuleBuilder.kt
@@ -5,6 +5,7 @@
 
 package org.jetbrains.kotlin.analysis.project.structure.builder
 
+import org.jetbrains.kotlin.analysis.api.KaExperimentalApi
 import org.jetbrains.kotlin.analysis.api.standalone.base.projectStructure.StandaloneProjectFactory
 import org.jetbrains.kotlin.analysis.api.projectStructure.KaLibraryModule
 import org.jetbrains.kotlin.analysis.api.projectStructure.KaLibrarySourceModule
@@ -23,9 +24,12 @@ public open class KtLibraryModuleBuilder(
 
     override fun build(): KaLibraryModule = build(isSdk = false)
 
+    @OptIn(KaExperimentalApi::class)
     protected fun build(isSdk: Boolean): KaLibraryModule {
         val binaryRoots = getBinaryRoots()
-        val contentScope = StandaloneProjectFactory.createSearchScopeByLibraryRoots(binaryRoots, kotlinCoreProjectEnvironment)
+        val binaryVirtualFiles = getBinaryVirtualFiles()
+        val contentScope =
+            StandaloneProjectFactory.createSearchScopeByLibraryRoots(binaryRoots, binaryVirtualFiles, kotlinCoreProjectEnvironment)
         return KaLibraryModuleImpl(
             directRegularDependencies,
             directDependsOnDependencies,
@@ -34,6 +38,7 @@ public open class KtLibraryModuleBuilder(
             platform,
             kotlinCoreProjectEnvironment.project,
             binaryRoots,
+            binaryVirtualFiles,
             libraryName,
             librarySources,
             isSdk,
diff --git a/analysis/analysis-api-standalone/src/org/jetbrains/kotlin/analysis/project/structure/impl/KaLibraryModuleImpl.kt b/analysis/analysis-api-standalone/src/org/jetbrains/kotlin/analysis/project/structure/impl/KaLibraryModuleImpl.kt
index 3f018810c15a..500d70615bf5 100644
--- a/analysis/analysis-api-standalone/src/org/jetbrains/kotlin/analysis/project/structure/impl/KaLibraryModuleImpl.kt
+++ b/analysis/analysis-api-standalone/src/org/jetbrains/kotlin/analysis/project/structure/impl/KaLibraryModuleImpl.kt
@@ -6,7 +6,9 @@
 package org.jetbrains.kotlin.analysis.project.structure.impl
 
 import com.intellij.openapi.project.Project
+import com.intellij.openapi.vfs.VirtualFile
 import com.intellij.psi.search.GlobalSearchScope
+import org.jetbrains.kotlin.analysis.api.KaExperimentalApi
 import org.jetbrains.kotlin.analysis.api.KaPlatformInterface
 import org.jetbrains.kotlin.analysis.api.platform.projectStructure.computeTransitiveDependsOnDependencies
 import org.jetbrains.kotlin.analysis.api.projectStructure.KaLibraryModule
@@ -15,6 +17,7 @@ import org.jetbrains.kotlin.analysis.api.projectStructure.KaModule
 import org.jetbrains.kotlin.platform.TargetPlatform
 import java.nio.file.Path
 
+@KaExperimentalApi
 internal class KaLibraryModuleImpl(
     override val directRegularDependencies: List<KaModule>,
     override val directDependsOnDependencies: List<KaModule>,
@@ -23,6 +26,7 @@ internal class KaLibraryModuleImpl(
     override val targetPlatform: TargetPlatform,
     override val project: Project,
     override val binaryRoots: Collection<Path>,
+    override val binaryVirtualFiles: Collection<VirtualFile>,
     override val libraryName: String,
     override val librarySources: KaLibrarySourceModule?,
 
diff --git a/analysis/analysis-api/src/org/jetbrains/kotlin/analysis/api/projectStructure/KaModule.kt b/analysis/analysis-api/src/org/jetbrains/kotlin/analysis/api/projectStructure/KaModule.kt
index f4d634e5d070..c9e8e0473ac2 100644
--- a/analysis/analysis-api/src/org/jetbrains/kotlin/analysis/api/projectStructure/KaModule.kt
+++ b/analysis/analysis-api/src/org/jetbrains/kotlin/analysis/api/projectStructure/KaModule.kt
@@ -8,6 +8,7 @@
 package org.jetbrains.kotlin.analysis.api.projectStructure
 
 import com.intellij.openapi.project.Project
+import com.intellij.openapi.vfs.VirtualFile
 import com.intellij.psi.PsiFile
 import com.intellij.psi.search.GlobalSearchScope
 import org.jetbrains.kotlin.analysis.api.KaExperimentalApi
@@ -140,6 +141,12 @@ public interface KaLibraryModule : KaModule {
      */
     public val binaryRoots: Collection<Path>
 
+    /**
+     * A list of binary files in [VirtualFile] form if the library module represents a library in in-memory file system.
+     */
+    @KaExperimentalApi
+    public val binaryVirtualFiles: Collection<VirtualFile>
+
     /**
      * A library source, if any. If current module is a binary JAR, then [librarySources] corresponds to the sources JAR.
      */
diff --git a/analysis/analysis-test-framework/tests/org/jetbrains/kotlin/analysis/test/framework/projectStructure/KaLibraryModuleImpl.kt b/analysis/analysis-test-framework/tests/org/jetbrains/kotlin/analysis/test/framework/projectStructure/KaLibraryModuleImpl.kt
index eb3204ef8293..cdf8c4e28783 100644
--- a/analysis/analysis-test-framework/tests/org/jetbrains/kotlin/analysis/test/framework/projectStructure/KaLibraryModuleImpl.kt
+++ b/analysis/analysis-test-framework/tests/org/jetbrains/kotlin/analysis/test/framework/projectStructure/KaLibraryModuleImpl.kt
@@ -6,6 +6,7 @@
 package org.jetbrains.kotlin.analysis.test.framework.projectStructure
 
 import com.intellij.openapi.project.Project
+import com.intellij.openapi.vfs.VirtualFile
 import com.intellij.psi.search.GlobalSearchScope
 import org.jetbrains.kotlin.analysis.api.projectStructure.KaLibraryModule
 import org.jetbrains.kotlin.analysis.api.projectStructure.KaLibrarySourceModule
@@ -26,5 +27,7 @@ class KaLibraryModuleImpl(
     override val directDependsOnDependencies: MutableList<KaModule> = mutableListOf()
     override val directFriendDependencies: MutableList<KaModule> = mutableListOf()
 
+    override val binaryVirtualFiles: Collection<VirtualFile> = emptyList()
+
     override fun toString(): String = libraryName
 }
diff --git a/analysis/analysis-test-framework/tests/org/jetbrains/kotlin/analysis/test/framework/projectStructure/KaSourceModuleByCompilerConfiguration.kt b/analysis/analysis-test-framework/tests/org/jetbrains/kotlin/analysis/test/framework/projectStructure/KaSourceModuleByCompilerConfiguration.kt
index 79a888af0c6b..68f9dac96f82 100644
--- a/analysis/analysis-test-framework/tests/org/jetbrains/kotlin/analysis/test/framework/projectStructure/KaSourceModuleByCompilerConfiguration.kt
+++ b/analysis/analysis-test-framework/tests/org/jetbrains/kotlin/analysis/test/framework/projectStructure/KaSourceModuleByCompilerConfiguration.kt
@@ -6,6 +6,7 @@
 package org.jetbrains.kotlin.analysis.test.framework.projectStructure
 
 import com.intellij.openapi.project.Project
+import com.intellij.openapi.vfs.VirtualFile
 import com.intellij.psi.PsiFile
 import com.intellij.psi.search.GlobalSearchScope
 import org.jetbrains.kotlin.analysis.api.platform.projectStructure.computeTransitiveDependsOnDependencies
@@ -73,6 +74,7 @@ abstract class KtModuleByCompilerConfiguration(
         val jdkHomePaths = StandaloneProjectFactory.getDefaultJdkModulePaths(project, jdkHome.toPath())
         val scope = StandaloneProjectFactory.createSearchScopeByLibraryRoots(
             jdkHomePaths,
+            emptyList(),
             testServices.environmentManager.getProjectEnvironment()
         )
 
@@ -148,6 +150,7 @@ class KaLibraryModuleByCompilerConfiguration(
     override val libraryName: String get() = testModule.name
     override val librarySources: KaLibrarySourceModule? get() = null
     override val isSdk: Boolean get() = false
+    override val binaryVirtualFiles: Collection<VirtualFile> = emptyList()
 
     override val contentScope: GlobalSearchScope =
         GlobalSearchScope.filesScope(project, psiFiles.map { it.virtualFile })
@@ -174,6 +177,7 @@ private class LibraryByRoots(
 ) : KaLibraryModule {
     override val contentScope: GlobalSearchScope = StandaloneProjectFactory.createSearchScopeByLibraryRoots(
         roots,
+        emptyList(),
         testServices.environmentManager.getProjectEnvironment(),
     )
     override val libraryName: String get() = "Test Library $roots"
@@ -184,6 +188,7 @@ private class LibraryByRoots(
     override val targetPlatform: TargetPlatform get() = parentModule.targetPlatform
     override val binaryRoots: Collection<Path> get() = roots
     override val isSdk: Boolean get() = false
+    override val binaryVirtualFiles: Collection<VirtualFile> = emptyList()
 
     override fun equals(other: Any?): Boolean {
         if (this === other) return true
diff --git a/analysis/analysis-test-framework/tests/org/jetbrains/kotlin/analysis/test/framework/projectStructure/KtLibraryBinaryTestModuleFactoryBase.kt b/analysis/analysis-test-framework/tests/org/jetbrains/kotlin/analysis/test/framework/projectStructure/KtLibraryBinaryTestModuleFactoryBase.kt
index 1fa92bb5adfd..1ec30ad13470 100644
--- a/analysis/analysis-test-framework/tests/org/jetbrains/kotlin/analysis/test/framework/projectStructure/KtLibraryBinaryTestModuleFactoryBase.kt
+++ b/analysis/analysis-test-framework/tests/org/jetbrains/kotlin/analysis/test/framework/projectStructure/KtLibraryBinaryTestModuleFactoryBase.kt
@@ -38,6 +38,7 @@ abstract class KtLibraryBinaryTestModuleFactoryBase : KtTestModuleFactory {
                 testModule.targetPlatform,
                 StandaloneProjectFactory.createSearchScopeByLibraryRoots(
                     listOf(binaryRoot),
+                    emptyList(),
                     testServices.environmentManager.getProjectEnvironment(),
                 ),
                 project,
diff --git a/analysis/analysis-test-framework/tests/org/jetbrains/kotlin/analysis/test/framework/projectStructure/KtLibrarySourceTestModuleFactory.kt b/analysis/analysis-test-framework/tests/org/jetbrains/kotlin/analysis/test/framework/projectStructure/KtLibrarySourceTestModuleFactory.kt
index 8c4723ee606e..7f3ef5ebf9e1 100644
--- a/analysis/analysis-test-framework/tests/org/jetbrains/kotlin/analysis/test/framework/projectStructure/KtLibrarySourceTestModuleFactory.kt
+++ b/analysis/analysis-test-framework/tests/org/jetbrains/kotlin/analysis/test/framework/projectStructure/KtLibrarySourceTestModuleFactory.kt
@@ -59,6 +59,7 @@ fun createKtLibrarySourceModule(
         testModule.targetPlatform,
         StandaloneProjectFactory.createSearchScopeByLibraryRoots(
             listOf(libraryJar),
+            emptyList(),
             testServices.environmentManager.getProjectEnvironment(),
         ),
         project,
diff --git a/analysis/analysis-test-framework/tests/org/jetbrains/kotlin/analysis/test/framework/projectStructure/TestModuleStructureFactory.kt b/analysis/analysis-test-framework/tests/org/jetbrains/kotlin/analysis/test/framework/projectStructure/TestModuleStructureFactory.kt
index d6866aa05b2b..f0d6c82f9dc0 100644
--- a/analysis/analysis-test-framework/tests/org/jetbrains/kotlin/analysis/test/framework/projectStructure/TestModuleStructureFactory.kt
+++ b/analysis/analysis-test-framework/tests/org/jetbrains/kotlin/analysis/test/framework/projectStructure/TestModuleStructureFactory.kt
@@ -219,6 +219,7 @@ object TestModuleStructureFactory {
     fun getScopeForLibraryByRoots(roots: Collection<Path>, testServices: TestServices): GlobalSearchScope {
         return StandaloneProjectFactory.createSearchScopeByLibraryRoots(
             roots,
+            emptyList(),
             testServices.environmentManager.getProjectEnvironment()
         )
     }

