diff --git a/android-lint-deps/BUILD.bazel b/android-lint-deps/BUILD.bazel
new file mode 100644
index 000000000000..8b387ce1bbef
--- /dev/null
+++ b/android-lint-deps/BUILD.bazel
@@ -0,0 +1,105 @@
+load("@rules_java//java:defs.bzl", "java_binary", "java_import")
+
+# ./bazel.cmd build //android-lint-deps:intellij-core_deploy.jar
+# ./bazel.cmd build //android-lint-deps:intellij-core_deploy-src.jar
+java_binary(
+    name = "intellij-core",
+    deploy_env = [":intellij-core-excluded-deps"],
+    main_class = "nil",
+    runtime_deps = [
+        "//java/java-psi-impl:psi-impl",
+        # "Shallow" modules are included but with transitive dependencies excluded.
+        # Eventually it might be safer to explicitly list the excluded deps.
+        ":java-indexing-impl-shallow",
+        ":jps-model-impl-shallow",
+        ":analysis-api-shallow",
+        ":analysis-impl-shallow",
+    ],
+)
+
+# ./bazel.cmd build //android-lint-deps:uast_deploy.jar
+# ./bazel.cmd build //android-lint-deps:uast_deploy-src.jar
+java_binary(
+    name = "uast",
+    main_class = "nil",
+    runtime_deps = [
+        ":uast-common-shallow",
+        ":uast-java-shallow",
+        ":uast-kotlin-base-shallow",
+        ":uast-kotlin-shallow",
+        ":uast-kotlin-fir-shallow",
+    ],
+)
+
+# These dependencies will be excluded from intellij-core.
+java_binary(
+    name = "intellij-core-excluded-deps",
+    main_class = "nil",
+    runtime_deps = [
+        # The following are packaged separately.
+        "@lib//:guava",
+        "@lib//:kotlin-reflect",
+        "@lib//:kotlin-stdlib",
+        # The following are not needed at all.
+        "@lib//:automaton",  # Adds a lot of toplevel .aut files we do not need.
+    ],
+)
+
+# We add java-indexing-impl because (1) Kotlinc needs it for JavaCoreApplicationEnvironment
+# and (2) Lint needs it for converting .class files into PSI stubs.
+java_import(
+    name = "java-indexing-impl-shallow",
+    jars = ["//java/java-indexing-impl:indexing-impl.jar"],
+    srcjar = "//java/java-indexing-impl:indexing-impl-sources.jar",
+)
+
+# We use JavaSdkUtil from jps-model-impl (ideally we wouldn't).
+java_import(
+    name = "jps-model-impl-shallow",
+    jars = ["//jps/model-impl:model-impl.jar"],
+    srcjar = "//jps/model-impl:model-impl-sources.jar",
+)
+
+# DCL depends on GeneratedParserUtilBase from analysis-impl, which itself needs analysis-api.
+# Eventually it would be nice to extract DCL support out into a smaller dependency.
+java_import(
+    name = "analysis-api-shallow",
+    jars = ["//platform/analysis-api:analysis.jar"],
+    srcjar = "//platform/analysis-api:analysis-sources.jar",
+)
+
+java_import(
+    name = "analysis-impl-shallow",
+    jars = ["//platform/analysis-impl:analysis-impl.jar"],
+    srcjar = "//platform/analysis-impl:analysis-impl-sources.jar",
+)
+
+java_import(
+    name = "uast-common-shallow",
+    jars = ["//uast/uast-common:uast.jar"],
+    srcjar = "//uast/uast-common:uast-sources.jar",
+)
+
+java_import(
+    name = "uast-java-shallow",
+    jars = ["//uast/uast-java:java-uast.jar"],
+    srcjar = "//uast/uast-java:java-uast-sources.jar",
+)
+
+java_import(
+    name = "uast-kotlin-base-shallow",
+    jars = ["//plugins/kotlin/uast/uast-kotlin-base:uast-kotlin-base.jar"],
+    srcjar = "//plugins/kotlin/uast/uast-kotlin-base:uast-kotlin-base-sources.jar",
+)
+
+java_import(
+    name = "uast-kotlin-shallow",
+    jars = ["//plugins/kotlin/uast/uast-kotlin:uast-kotlin.jar"],
+    srcjar = "//plugins/kotlin/uast/uast-kotlin:uast-kotlin-sources.jar",
+)
+
+java_import(
+    name = "uast-kotlin-fir-shallow",
+    jars = ["//plugins/kotlin/uast/uast-kotlin-fir:kotlin-uast-uast-kotlin-k2.jar"],
+    srcjar = "//plugins/kotlin/uast/uast-kotlin-fir:kotlin-uast-uast-kotlin-k2-sources.jar",
+)
