diff --git a/platform/analysis-api/BUILD.bazel b/platform/analysis-api/BUILD.bazel
index 29e8a18aaefc..5900d69c6d18 100644
--- a/platform/analysis-api/BUILD.bazel
+++ b/platform/analysis-api/BUILD.bazel
@@ -12,6 +12,7 @@ jvm_library(
   module_name = "intellij.platform.analysis",
   visibility = ["//visibility:public"],
   srcs = glob(["src/**/*.kt", "src/**/*.java", "src/**/*.form"], allow_empty = True),
+  kotlinc_opts = "@community//:k17",
   resources = [":analysis_resources"],
   deps = [
     "//jps/model-api:model",
diff --git a/platform/analysis-api/src/com/intellij/codeInspection/ex/QuickFixWrapper.java b/platform/analysis-api/src/com/intellij/codeInspection/ex/QuickFixWrapper.java
index e17e0452dfa7..5c3281bd314d 100644
--- a/platform/analysis-api/src/com/intellij/codeInspection/ex/QuickFixWrapper.java
+++ b/platform/analysis-api/src/com/intellij/codeInspection/ex/QuickFixWrapper.java
@@ -269,8 +269,7 @@ public final class QuickFixWrapper implements IntentionAction, PriorityAction, C
         if (myDescriptor.getStartElement() == null) return null;
         ActionContext descriptorContext = ActionContext.from(myDescriptor);
         return myUnwrappedAction.getPresentation(
-          descriptorContext.withOffset(Math.clamp(context.offset(), descriptorContext.selection().getStartOffset(),
-                                                  descriptorContext.selection().getEndOffset())));
+          descriptorContext.withOffset(Math.max(descriptorContext.selection().getStartOffset(), Math.min(descriptorContext.selection().getEndOffset(), context.offset()))));
       }
       PsiElement psiElement = myDescriptor.getPsiElement();
       if (psiElement == null || !psiElement.isValid()) return null;
diff --git a/platform/analysis-api/src/com/intellij/modcommand/ModCommand.java b/platform/analysis-api/src/com/intellij/modcommand/ModCommand.java
index 24f43753823a..2e3cd1ff01a3 100644
--- a/platform/analysis-api/src/com/intellij/modcommand/ModCommand.java
+++ b/platform/analysis-api/src/com/intellij/modcommand/ModCommand.java
@@ -67,7 +67,7 @@ public sealed interface ModCommand
     if (next.isEmpty()) return this;
     List<ModCommand> commands = new ArrayList<>(unpack());
     commands.addAll(next.unpack());
-    return commands.size() == 1 ? commands.get(0) : new ModCompositeCommand(commands); 
+    return commands.size() == 1 ? commands.get(0) : new ModCompositeCommand(commands);
   }
 
   /**
@@ -149,7 +149,7 @@ public sealed interface ModCommand
   }
 
   /**
-   * @param attributes attributes to use for highlighting 
+   * @param attributes attributes to use for highlighting
    * @param elements elements to highlight
    * @return a command to highlight the elements, assuming that nothing will be changed in the file
    */
@@ -206,7 +206,7 @@ public sealed interface ModCommand
   @ApiStatus.Experimental
   static @NotNull ModCommand updateOptionList(
     @NotNull PsiElement context, @NotNull @NonNls String bindId, @NotNull Consumer<@NotNull List<@NotNull String>> listUpdater) {
-    @SuppressWarnings("unchecked") 
+    @SuppressWarnings("unchecked")
     List<String> oldValue = (List<String>)Objects.requireNonNull(OptionControllerProvider.getOption(context, bindId));
     List<String> newValue = new ArrayList<>(oldValue);
     listUpdater.accept(newValue);
@@ -339,7 +339,7 @@ public sealed interface ModCommand
   /**
    * @param command a command to tune
    * @param file a file where we want to navigate
-   * @param offset an offset in the file before the command is executed 
+   * @param offset an offset in the file before the command is executed
    * @param leanRight if true, lean to the right side when the text was inserted right at the caret position
    * @return an updated command which tries to navigate inside the specified file, taking into account the modifications inside that file
    */
@@ -351,12 +351,12 @@ public sealed interface ModCommand
       if (sub instanceof ModUpdateFileText updateFileText && updateFileText.file().equals(virtualFile)) {
         offset = updateFileText.translateOffset(offset, leanRight);
       }
-      if (sub instanceof ModDeleteFile(VirtualFile fileToDelete) && fileToDelete.equals(virtualFile)) {
+      if (sub instanceof ModDeleteFile deleteFile && deleteFile.file().equals(virtualFile)) {
         // Navigation is useless: we are removing the target file
         return command;
       }
-      if (sub instanceof ModMoveFile(VirtualFile fileToMove, FutureVirtualFile targetFile) && fileToMove.equals(virtualFile)) {
-        virtualFile = targetFile;
+      if (sub instanceof ModMoveFile moveFile && moveFile.file().equals(virtualFile)) {
+        virtualFile = moveFile.targetFile();
       }
       if (!(sub instanceof ModNavigate)) {
         finalCommand = finalCommand.andThen(sub);
@@ -381,7 +381,7 @@ public sealed interface ModCommand
   }
 
   /**
-   * Creates a command that allows user to select arbitrary number of members (but at least one). 
+   * Creates a command that allows user to select arbitrary number of members (but at least one).
    * In batch mode, it's assumed that the default selection is selected.
    *
    * @param title user-readable title to display in UI
@@ -395,9 +395,9 @@ public sealed interface ModCommand
                                                    @NotNull List<? extends OptMultiSelector.@NotNull OptElement> defaultSelection,
                                                    @NotNull Function<@NotNull List<? extends OptMultiSelector.@NotNull OptElement>, ? extends @NotNull ModCommand> nextCommand) {
     class MemberHolder implements OptionContainer {
-      @SuppressWarnings("FieldMayBeFinal") 
+      @SuppressWarnings("FieldMayBeFinal")
       @NotNull List<? extends OptMultiSelector.@NotNull OptElement> myElements = new ArrayList<>(defaultSelection);
-      
+
       @Override
       public @NotNull OptPane getOptionsPane() {
         return OptPane.pane(
@@ -438,7 +438,7 @@ public sealed interface ModCommand
    * In batch mode, the first option is also selected automatically.
    *
    * @param title title to display to the user
-   * @param actions actions to select from. If there's only one action, then it could be executed right away without asking the user. 
+   * @param actions actions to select from. If there's only one action, then it could be executed right away without asking the user.
    * @see #psiUpdateStep(PsiElement, String, BiConsumer) could be useful as a subsequent step
    */
   static @NotNull ModCommand chooseAction(@NotNull @NlsContexts.PopupTitle String title,
@@ -448,7 +448,7 @@ public sealed interface ModCommand
 
   /**
    * Creates a command to move a file to a specified directory
-   * 
+   *
    * @param file file to move
    * @param target target directory
    * @return a command that moves the file to a specified directory
diff --git a/platform/analysis-api/src/com/intellij/modcommand/ModUpdateSystemOptions.java b/platform/analysis-api/src/com/intellij/modcommand/ModUpdateSystemOptions.java
index 8a6f98ca8660..03ad9a5b1713 100644
--- a/platform/analysis-api/src/com/intellij/modcommand/ModUpdateSystemOptions.java
+++ b/platform/analysis-api/src/com/intellij/modcommand/ModUpdateSystemOptions.java
@@ -10,8 +10,8 @@ import java.util.List;
 
 /**
  * A command that updates system options accessible via {@link com.intellij.codeInspection.options.OptionControllerProvider}.
- * Could be used, for example, to update inspection settings. 
- * 
+ * Could be used, for example, to update inspection settings.
+ *
  * @param options list of options to update
  */
 @ApiStatus.Experimental
@@ -23,7 +23,7 @@ public record ModUpdateSystemOptions(@NotNull List<@NotNull ModifiedOption> opti
 
   /**
    * A modified option record. Supported value types are boolean, int, long, double, String, enum, List&lt;String&gt;
-   * 
+   *
    * @param bindId id of the option. Should be accessible via {@link com.intellij.codeInspection.options.OptionControllerProvider}
    * @param oldValue old value of the option
    * @param newValue new value of the option
@@ -32,8 +32,8 @@ public record ModUpdateSystemOptions(@NotNull List<@NotNull ModifiedOption> opti
 
   @Override
   public @NotNull ModCommand andThen(@NotNull ModCommand next) {
-    if (next instanceof ModUpdateSystemOptions(var nextOptions)) {
-      return new ModUpdateSystemOptions(ContainerUtil.concat(options, nextOptions));
+    if (next instanceof ModUpdateSystemOptions nextCmd) {
+      return new ModUpdateSystemOptions(ContainerUtil.concat(options, nextCmd.options()));
     }
     return ModCommand.super.andThen(next);
   }
diff --git a/platform/analysis-impl/BUILD.bazel b/platform/analysis-impl/BUILD.bazel
index fca59b61bada..cd0cbf942b8d 100644
--- a/platform/analysis-impl/BUILD.bazel
+++ b/platform/analysis-impl/BUILD.bazel
@@ -12,6 +12,7 @@ jvm_library(
   module_name = "intellij.platform.analysis.impl",
   visibility = ["//visibility:public"],
   srcs = glob(["src/**/*.kt", "src/**/*.java", "src/**/*.form", "gen/**/*.kt", "gen/**/*.java"], allow_empty = True),
+  kotlinc_opts = "@community//:k17",
   resources = [":analysis-impl_resources"],
   deps = [
     "//platform/analysis-api:analysis",
diff --git a/platform/analysis-impl/src/com/intellij/codeInsight/completion/BaseCompletionLookupArranger.java b/platform/analysis-impl/src/com/intellij/codeInsight/completion/BaseCompletionLookupArranger.java
index 174aabd3bb31..2ad46a514531 100644
--- a/platform/analysis-impl/src/com/intellij/codeInsight/completion/BaseCompletionLookupArranger.java
+++ b/platform/analysis-impl/src/com/intellij/codeInsight/completion/BaseCompletionLookupArranger.java
@@ -141,7 +141,7 @@ public class BaseCompletionLookupArranger extends LookupArranger implements Comp
     if (pairs.isEmpty()) return true;
 
     for (int i = 1; i < pairs.size(); i++) {
-      if (!Comparing.equal(pairs.get(i).second, pairs.getFirst().second)) {
+      if (!Comparing.equal(pairs.get(i).second, pairs.get(0).second)) {
         return false;
       }
     }
@@ -629,7 +629,7 @@ public class BaseCompletionLookupArranger extends LookupArranger implements Comp
     if (exactMatches.isEmpty()) return null;
 
     if (exactMatches.size() == 1) {
-      return exactMatches.getFirst();
+      return exactMatches.get(0);
     }
 
     return sortByRelevance(groupItemsBySorter(exactMatches)).iterator().next();
diff --git a/platform/analysis-impl/src/com/intellij/codeInsight/daemon/impl/Divider.java b/platform/analysis-impl/src/com/intellij/codeInsight/daemon/impl/Divider.java
index b41b4c206688..f4bc547d6cc8 100644
--- a/platform/analysis-impl/src/com/intellij/codeInsight/daemon/impl/Divider.java
+++ b/platform/analysis-impl/src/com/intellij/codeInsight/daemon/impl/Divider.java
@@ -161,8 +161,8 @@ public final class Divider {
       }
     }
 
-    PsiElement parent = !outside.isEmpty() ? outside.getLast() :
-                        !inside.isEmpty() ? inside.getLast() :
+    PsiElement parent = !outside.isEmpty() ? outside.get(outside.size() - 1) :
+                        !inside.isEmpty() ? inside.get(inside.size() - 1) :
                         CollectHighlightsUtil.findCommonParent(root, startOffset, endOffset);
     while (parent != null && !(parent instanceof PsiFile)) {
       parent = parent.getParent();
diff --git a/platform/analysis-impl/src/com/intellij/codeInsight/daemon/impl/HighlightInfo.java b/platform/analysis-impl/src/com/intellij/codeInsight/daemon/impl/HighlightInfo.java
index ed8c3f0b2be5..5c6ac0f078c9 100644
--- a/platform/analysis-impl/src/com/intellij/codeInsight/daemon/impl/HighlightInfo.java
+++ b/platform/analysis-impl/src/com/intellij/codeInsight/daemon/impl/HighlightInfo.java
@@ -326,22 +326,24 @@ public class HighlightInfo implements Segment {
 
   @ApiStatus.Internal
   public final void addFileLevelComponent(@NotNull FileEditor fileEditor, @NotNull JComponent component) {
-    switch (fileLevelComponentsStorage) {
-      case null ->
-        fileLevelComponentsStorage = new Pair<>(fileEditor, component);
-      case Pair<?, ?> p -> {
-        //noinspection unchecked
-        Pair<FileEditor, JComponent> pair = (Pair<FileEditor, JComponent>)p;
-        Map<FileEditor, JComponent> map = new HashMap<>();
-        map.put(pair.first, pair.second);
-        map.put(fileEditor, component);
-        fileLevelComponentsStorage = map;
-      }
-      case Map<?, ?> map ->
-        //noinspection unchecked
-        ((Map<FileEditor, JComponent>)map).put(fileEditor, component);
-      default ->
-        LOG.error(new IllegalStateException("fileLevelComponents=" + fileLevelComponentsStorage));
+    Object storage = fileLevelComponentsStorage;
+    if (storage == null) {
+      fileLevelComponentsStorage = new Pair<>(fileEditor, component);
+    }
+    else if (storage instanceof Pair<?, ?>) {
+      //noinspection unchecked
+      Pair<FileEditor, JComponent> pair = (Pair<FileEditor, JComponent>)storage;
+      Map<FileEditor, JComponent> map = new HashMap<>();
+      map.put(pair.first, pair.second);
+      map.put(fileEditor, component);
+      fileLevelComponentsStorage = map;
+    }
+    else if (storage instanceof Map<?, ?>) {
+      //noinspection unchecked
+      ((Map<FileEditor, JComponent>)storage).put(fileEditor, component);
+    }
+    else {
+      LOG.error(new IllegalStateException("fileLevelComponents=" + storage));
     }
   }
 
@@ -362,21 +364,20 @@ public class HighlightInfo implements Segment {
 
   @ApiStatus.Internal
   public final @Nullable JComponent getFileLevelComponent(@NotNull FileEditor fileEditor) {
-    switch (fileLevelComponentsStorage) {
-      case null -> {
-        return null;
-      }
-      case Pair<?,?> pair -> {
-        return pair.first == fileEditor ? (JComponent)pair.second : null;
-      }
-      case Map<?, ?> map -> {
-        //noinspection unchecked
-        return ((Map<FileEditor, JComponent>)map).get(fileEditor);
-      }
-      default -> {
-        LOG.error(new IllegalStateException("fileLevelComponents=" + fileLevelComponentsStorage));
-        return null;
-      }
+    Object storage = fileLevelComponentsStorage;
+    if (storage == null) {
+      return null;
+    }
+    else if (storage instanceof Pair<?, ?> pair) {
+      return pair.first == fileEditor ? (JComponent)pair.second : null;
+    }
+    else if (storage instanceof Map<?, ?> map) {
+      //noinspection unchecked
+      return ((Map<FileEditor, JComponent>)map).get(fileEditor);
+    }
+    else {
+      LOG.error(new IllegalStateException("fileLevelComponents=" + storage));
+      return null;
     }
   }
 
diff --git a/platform/analysis-impl/src/com/intellij/lang/impl/modcommand/ModCommandBatchExecutorImpl.java b/platform/analysis-impl/src/com/intellij/lang/impl/modcommand/ModCommandBatchExecutorImpl.java
index 67b68391308e..9da860d3f3b2 100644
--- a/platform/analysis-impl/src/com/intellij/lang/impl/modcommand/ModCommandBatchExecutorImpl.java
+++ b/platform/analysis-impl/src/com/intellij/lang/impl/modcommand/ModCommandBatchExecutorImpl.java
@@ -42,16 +42,16 @@ import static com.intellij.openapi.util.text.HtmlChunk.tag;
 import static com.intellij.openapi.util.text.HtmlChunk.text;
 
 /**
- * An {@link ModCommandExecutor} service implementation that supports batch execution only. 
- * Interactive execution request executes like in batch. 
+ * An {@link ModCommandExecutor} service implementation that supports batch execution only.
+ * Interactive execution request executes like in batch.
  */
 @ApiStatus.Internal
 public class ModCommandBatchExecutorImpl implements ModCommandExecutor {
   /**
    * {@inheritDoc}
-   * 
+   *
    * @implNote this implementation executes the action like in batch, without displaying any UI.
-   * 
+   *
    * @param context current context
    * @param command a command to execute
    * @param editor ignored by this implementation, can be null
@@ -86,37 +86,40 @@ public class ModCommandBatchExecutorImpl implements ModCommandExecutor {
         command instanceof ModUpdateReferences || command instanceof ModOpenUrl) {
       return Result.INTERACTIVE;
     }
-    return switch (command) {
-      case ModUpdateFileText upd -> executeUpdate(project, upd) ? Result.SUCCESS : Result.ABORT;
-      case ModCreateFile create -> {
-        String message = executeCreate(project, create);
-        yield message == null ? Result.SUCCESS : new Error(message);
-      }
-      case ModDeleteFile deleteFile -> {
-        String message = executeDelete(deleteFile);
-        yield message == null ? Result.SUCCESS : new Error(message);
-      }
-      case ModMoveFile moveFile -> {
-        String message = executeMove(moveFile);
-        yield message == null ? Result.SUCCESS : new Error(message);
-      }
-      case ModCompositeCommand(var commands) -> {
-        BatchExecutionResult result = Result.NOTHING;
-        for (ModCommand subCommand : commands) {
-          result = result.compose(doExecuteInBatch(context, subCommand));
-          if (result == Result.ABORT || result instanceof Error) break;
-        }
-        yield result;
+    if (command instanceof ModUpdateFileText upd) return executeUpdate(project, upd) ? Result.SUCCESS : Result.ABORT;
+    if (command instanceof ModCreateFile create) {
+      String message = executeCreate(project, create);
+      return message == null ? Result.SUCCESS : new Error(message);
+    }
+    if (command instanceof ModDeleteFile deleteFile) {
+      String message = executeDelete(deleteFile);
+      return message == null ? Result.SUCCESS : new Error(message);
+    }
+    if (command instanceof ModMoveFile moveFile) {
+      String message = executeMove(moveFile);
+      return message == null ? Result.SUCCESS : new Error(message);
+    }
+    if (command instanceof ModCompositeCommand composite) {
+      List<ModCommand> commands = composite.commands();
+      BatchExecutionResult result = Result.NOTHING;
+      for (ModCommand subCommand : commands) {
+        result = result.compose(doExecuteInBatch(context, subCommand));
+        if (result == Result.ABORT || result instanceof Error) break;
       }
-      case ModChooseAction chooser -> executeChooseInBatch(context, chooser);
-      case ModShowConflicts ignored -> Result.CONFLICTS;
-      case ModEditOptions<?> editOptions -> bypassEditOptions(editOptions, context);
-      case ModDisplayMessage(String text, var kind) -> switch (kind) {
+      return result;
+    }
+    if (command instanceof ModChooseAction chooser) return executeChooseInBatch(context, chooser);
+    if (command instanceof ModShowConflicts) return Result.CONFLICTS;
+    if (command instanceof ModEditOptions<?> editOptions) return bypassEditOptions(editOptions, context);
+    if (command instanceof ModDisplayMessage displayMessage) {
+      String text = displayMessage.messageText();
+      ModDisplayMessage.MessageKind kind = displayMessage.kind();
+      return switch (kind) {
         case ERROR -> new Error(text);
         case INFORMATION -> Result.INTERACTIVE;
       };
-      default -> throw new IllegalArgumentException("Unknown command: " + command);
-    };
+    }
+    throw new IllegalArgumentException("Unknown command: " + command);
   }
 
   private <T extends OptionContainer> BatchExecutionResult bypassEditOptions(@NotNull ModEditOptions<T> options, @NotNull ActionContext context) {
@@ -175,7 +178,7 @@ public class ModCommandBatchExecutorImpl implements ModCommandExecutor {
               source.move(this, origParent);
             }
             catch (IOException ignored) {
-              // Ignore move exception 
+              // Ignore move exception
             }
           }
           return AnalysisBundle.message("modcommand.executor.cannot.rename.file",
@@ -196,14 +199,16 @@ public class ModCommandBatchExecutorImpl implements ModCommandExecutor {
           return null;
         }
         VirtualFile newFile = parent.createChildData(this, file.getName());
-        if (create.content() instanceof ModCreateFile.Text(String text)) {
+        if (create.content() instanceof ModCreateFile.Text textContent) {
+          String text = textContent.text();
           PsiFile psiFile = PsiManager.getInstance(project).findFile(newFile);
           if (psiFile == null) return AnalysisBundle.message("modcommand.executor.unable.to.find.the.new.file", file.getName());
           Document document = psiFile.getViewProvider().getDocument();
           document.setText(text);
           PsiDocumentManager.getInstance(project).commitDocument(document);
         }
-        else if (create.content() instanceof ModCreateFile.Binary(byte[] bytes)) {
+        else if (create.content() instanceof ModCreateFile.Binary binaryContent) {
+          byte[] bytes = binaryContent.bytes();
           newFile.setBinaryContent(bytes);
         }
         return null;
@@ -321,12 +326,14 @@ public class ModCommandBatchExecutorImpl implements ModCommandExecutor {
                                                                modFile.newText(),
                                                                true));
       }
-      else if (command instanceof ModCreateFile(FutureVirtualFile createdFile, ModCreateFile.Content content)) {
+      else if (command instanceof ModCreateFile createFile) {
+        FutureVirtualFile createdFile = createFile.file();
+        ModCreateFile.Content content = createFile.content();
         if (content instanceof ModCreateFile.Directory) {
           createdDirs.add(getFileNamePresentation(project, createdFile));
         } else {
           String contentRepresentation =
-            content instanceof ModCreateFile.Text(String text) ? text : AnalysisBundle.message("preview.binary.content");
+            content instanceof ModCreateFile.Text textContent ? textContent.text() : AnalysisBundle.message("preview.binary.content");
           customDiffList.add(new IntentionPreviewInfo.CustomDiff(createdFile.getFileType(),
                                                                  getFileNamePresentation(project, createdFile),
                                                                  "",
@@ -349,7 +356,9 @@ public class ModCommandBatchExecutorImpl implements ModCommandExecutor {
       else if (command instanceof ModEditOptions<?> target) {
         return getEditOptionsPreview(context, target);
       }
-      else if (command instanceof ModDisplayMessage(String text, ModDisplayMessage.MessageKind kind)) {
+      else if (command instanceof ModDisplayMessage displayMessage) {
+        String text = displayMessage.messageText();
+        ModDisplayMessage.MessageKind kind = displayMessage.kind();
         if (kind == ModDisplayMessage.MessageKind.ERROR) {
           return new IntentionPreviewInfo.Html(new HtmlBuilder().append(
             AnalysisBundle.message("preview.cannot.perform.action")).br().append(text).toFragment(), IntentionPreviewInfo.InfoKind.ERROR);
@@ -358,15 +367,19 @@ public class ModCommandBatchExecutorImpl implements ModCommandExecutor {
           navigateInfo = new IntentionPreviewInfo.Html(text);
         }
       }
-      else if (command instanceof ModCopyToClipboard(String content)) {
+      else if (command instanceof ModCopyToClipboard copyToClipboard) {
+        String content = copyToClipboard.content();
         navigateInfo = new IntentionPreviewInfo.Html(text(
           AnalysisBundle.message("preview.copy.to.clipboard", StringUtil.shortenTextWithEllipsis(content, 50, 10))));
       }
-      else if (command instanceof ModOpenUrl(String url)) {
+      else if (command instanceof ModOpenUrl openUrl) {
+        String url = openUrl.url();
         navigateInfo = new IntentionPreviewInfo.Html(text(
           AnalysisBundle.message("preview.open.url", StringUtil.shortenTextWithEllipsis(url, 50, 10))));
       }
-      else if (command instanceof ModMoveFile(VirtualFile fileToCreate, FutureVirtualFile targetFile)) {
+      else if (command instanceof ModMoveFile moveFile) {
+        VirtualFile fileToCreate = moveFile.file();
+        FutureVirtualFile targetFile = moveFile.targetFile();
         IntentionPreviewInfo.Html html;
         if (targetFile.getName().equals(fileToCreate.getName())) {
           html = (IntentionPreviewInfo.Html)IntentionPreviewInfo.moveToDirectory(fileToCreate, targetFile.getParent());
diff --git a/platform/analysis-impl/src/com/intellij/profile/codeInspection/ProjectInspectionProfileManager.kt b/platform/analysis-impl/src/com/intellij/profile/codeInspection/ProjectInspectionProfileManager.kt
index 7cacd5869f9e..38c806e0745c 100644
--- a/platform/analysis-impl/src/com/intellij/profile/codeInspection/ProjectInspectionProfileManager.kt
+++ b/platform/analysis-impl/src/com/intellij/profile/codeInspection/ProjectInspectionProfileManager.kt
@@ -4,7 +4,7 @@ package com.intellij.profile.codeInspection
 import com.intellij.codeInspection.ex.InspectionProfileImpl
 import com.intellij.codeInspection.ex.ProjectInspectionToolRegistrar
 import com.intellij.configurationStore.*
-import com.intellij.diagnostic.runActivity
+
 import com.intellij.openapi.Disposable
 import com.intellij.openapi.application.ApplicationManager
 import com.intellij.openapi.components.*
@@ -94,10 +94,8 @@ open class ProjectInspectionProfileManager(final override val project: Project)
       return
     }
 
-    runActivity("project inspection profile loading") {
-      schemeManager.loadSchemes()
-      currentProfile.initInspectionTools(project)
-    }
+    schemeManager.loadSchemes()
+    currentProfile.initInspectionTools(project)
 
     StartupManager.getInstance(project).runAfterOpened {
       project.messageBus.syncPublisher(ProfileChangeAdapter.TOPIC).profilesInitialized()
diff --git a/java/java-indexing-impl/BUILD.bazel b/java/java-indexing-impl/BUILD.bazel
index 775d2eff9753..27cbc7d71864 100644
--- a/java/java-indexing-impl/BUILD.bazel
+++ b/java/java-indexing-impl/BUILD.bazel
@@ -12,6 +12,7 @@ jvm_library(
   module_name = "intellij.java.indexing.impl",
   visibility = ["//visibility:public"],
   srcs = glob(["src/**/*.kt", "src/**/*.java", "src/**/*.form"], allow_empty = True),
+  kotlinc_opts = "@community//:k17",
   resources = [":indexing-impl_resources"],
   deps = [
     "//java/java-psi-api:psi",
diff --git a/platform/plugins/parser/impl/src/com/intellij/platform/plugins/parser/impl/PluginDescriptorFromXmlStreamConsumer.kt b/platform/plugins/parser/impl/src/com/intellij/platform/plugins/parser/impl/PluginDescriptorFromXmlStreamConsumer.kt
index c959d84465e4..9536080969c3 100644
--- a/platform/plugins/parser/impl/src/com/intellij/platform/plugins/parser/impl/PluginDescriptorFromXmlStreamConsumer.kt
+++ b/platform/plugins/parser/impl/src/com/intellij/platform/plugins/parser/impl/PluginDescriptorFromXmlStreamConsumer.kt
@@ -40,7 +40,7 @@ class PluginDescriptorFromXmlStreamConsumer private constructor(
   }
 
   internal fun popIncludeBase() {
-    includeBaseStack.removeLast()
+    includeBaseStack.removeAt(includeBaseStack.lastIndex)
   }
 
   companion object {
