diff --git a/compiler/light-classes/src/org/jetbrains/kotlin/asJava/classes/KtLightClassForFacadeImpl.kt b/compiler/light-classes/src/org/jetbrains/kotlin/asJava/classes/KtLightClassForFacadeImpl.kt
index 60c0ec74421..b0d2f01b672 100644
--- a/compiler/light-classes/src/org/jetbrains/kotlin/asJava/classes/KtLightClassForFacadeImpl.kt
+++ b/compiler/light-classes/src/org/jetbrains/kotlin/asJava/classes/KtLightClassForFacadeImpl.kt
@@ -233,7 +233,7 @@ open class KtLightClassForFacadeImpl constructor(
         fun createForFacadeNoCache(fqName: FqName, searchScope: GlobalSearchScope, project: Project): KtLightClassForFacade? {
 
             val sources = KotlinAsJavaSupport.getInstance(project).findFilesForFacade(fqName, searchScope)
-                .filterNot { it.isCompiled }
+                .filterNot { it.isCompiled }.filterNot { it.isScript() }
 
             if (sources.isEmpty()) return null
 
@@ -245,7 +245,7 @@ open class KtLightClassForFacadeImpl constructor(
 
             return LightClassGenerationSupport.getInstance(project).run {
                 if (useUltraLightClasses) createUltraLightClassForFacade(manager, fqName, stubValue, sources)
-                    ?: error { "Unable to create UL class for facade" }
+                    ?: error("Unable to create UL class for facade")
                 else KtLightClassForFacadeImpl(manager, fqName, stubValue, sources)
             }
         }
