diff --git a/plugins/kotlin/uast/uast-kotlin-fir/src/org/jetbrains/uast/kotlin/FirKotlinUastResolveProviderService.kt b/plugins/kotlin/uast/uast-kotlin-fir/src/org/jetbrains/uast/kotlin/FirKotlinUastResolveProviderService.kt
index c6adc0b3d2b7..7650bd769f82 100644
--- a/plugins/kotlin/uast/uast-kotlin-fir/src/org/jetbrains/uast/kotlin/FirKotlinUastResolveProviderService.kt
+++ b/plugins/kotlin/uast/uast-kotlin-fir/src/org/jetbrains/uast/kotlin/FirKotlinUastResolveProviderService.kt
@@ -229,7 +229,7 @@ interface FirKotlinUastResolveProviderService : BaseKotlinUastResolveProviderSer
                 val psi = analyzeForUast(ktExpression) {
                     val candidate = candidatePointer.restoreSymbol() ?: return@forEach
                     when (candidate) {
-                        is KaVariableSymbol -> psiForUast(candidate)
+                        is KaVariableSymbol -> psiForUast(candidate, ktExpression)
                         is KaFunctionSymbol -> toPsiMethod(candidate, ktExpression)
                     }
                 }
@@ -455,7 +455,7 @@ interface FirKotlinUastResolveProviderService : BaseKotlinUastResolveProviderSer
                         resolvedTargetSymbol.owningProperty.psi
                     }
                     else -> {
-                        psiForUast(resolvedTargetSymbol)
+                        psiForUast(resolvedTargetSymbol, ktExpression)
                     }
                 }
 
diff --git a/plugins/kotlin/uast/uast-kotlin-fir/src/org/jetbrains/uast/kotlin/internal/firKotlinInternalUastUtils.kt b/plugins/kotlin/uast/uast-kotlin-fir/src/org/jetbrains/uast/kotlin/internal/firKotlinInternalUastUtils.kt
index 1a5127b998ac..4b33309ec236 100644
--- a/plugins/kotlin/uast/uast-kotlin-fir/src/org/jetbrains/uast/kotlin/internal/firKotlinInternalUastUtils.kt
+++ b/plugins/kotlin/uast/uast-kotlin-fir/src/org/jetbrains/uast/kotlin/internal/firKotlinInternalUastUtils.kt
@@ -88,20 +88,11 @@ internal fun toPsiClass(
 
 context(KaSession)
 @OptIn(KaExperimentalApi::class)
-internal fun toPsiMethod(
+private fun fakePsiMethodForReifiedInline(
     functionSymbol: KaFunctionSymbol,
     context: KtElement,
     kaCallInfo: KaCallInfo? = null,
 ): PsiMethod? {
-    // Error handling for a case like KTIJ-23503: Outer.<no name provided>.Inner from broken code
-    val nameToCheck = if (functionSymbol is KaConstructorSymbol)
-        functionSymbol.containingClassId?.asSingleFqName()
-    else
-        functionSymbol.callableId?.asSingleFqName()
-    if (nameToCheck?.pathSegments()?.any { it.isSpecial } == true) {
-        return null
-    }
-
     // `inline` w/ `reified` type param from binary dependency,
     // which we can't find source PSI, so fake it
     if (functionSymbol.origin == KaSymbolOrigin.LIBRARY &&
@@ -122,7 +113,27 @@ internal fun toPsiMethod(
                 }
         }
     }
-    return when (val psi = psiForUast(functionSymbol)) {
+    return null
+}
+
+context(KaSession)
+internal fun toPsiMethod(
+    functionSymbol: KaFunctionSymbol,
+    context: KtElement,
+    kaCallInfo: KaCallInfo? = null,
+): PsiMethod? {
+    // Error handling for a case like KTIJ-23503: Outer.<no name provided>.Inner from broken code
+    val nameToCheck = if (functionSymbol is KaConstructorSymbol)
+        functionSymbol.containingClassId?.asSingleFqName()
+    else
+        functionSymbol.callableId?.asSingleFqName()
+    if (nameToCheck?.pathSegments()?.any { it.isSpecial } == true) {
+        return null
+    }
+
+    fakePsiMethodForReifiedInline(functionSymbol, context, kaCallInfo)?.let { return it }
+
+    return when (val psi = psiForUast(functionSymbol, context)) {
         null -> {
             // Lint/UAST CLI: try `fake` creation for a deserialized declaration
             toPsiMethodForDeserialized(functionSymbol, context, psi, kaCallInfo)
@@ -443,19 +454,26 @@ internal fun getKtType(ktCallableDeclaration: KtCallableDeclaration): KaType? {
 }
 
 /**
- * Finds Java stub-based [PsiElement] for symbols that refer to declarations in [KaLibraryModule].
+ * Finds Java stub-based [PsiElement] for symbols that refer to declarations from [KaSymbolOrigin.LIBRARY]
  */
 context(KaSession)
 @OptIn(KaExperimentalApi::class)
-internal tailrec fun psiForUast(symbol: KaSymbol): PsiElement? {
+internal tailrec fun psiForUast(
+    symbol: KaSymbol,
+    context: KtElement,
+): PsiElement? {
     if (symbol.origin == KaSymbolOrigin.LIBRARY) {
+        if (symbol is KaFunctionSymbol) {
+            fakePsiMethodForReifiedInline(symbol, context, null)?.let { return it }
+        }
+
         val psiProvider = FirKotlinUastLibraryPsiProviderService.getInstance()
         return with(psiProvider) { provide(symbol) }
     }
 
     if (symbol is KaConstructorSymbol) {
         symbol.originalConstructorIfTypeAliased?.let { originalConstructorSymbol ->
-            return psiForUast(originalConstructorSymbol)
+            return psiForUast(originalConstructorSymbol, context)
         }
     }
 
@@ -463,7 +481,7 @@ internal tailrec fun psiForUast(symbol: KaSymbol): PsiElement? {
         if (symbol.origin == KaSymbolOrigin.INTERSECTION_OVERRIDE || symbol.origin == KaSymbolOrigin.SUBSTITUTION_OVERRIDE) {
             val originalSymbol = symbol.fakeOverrideOriginal
             if (originalSymbol != symbol) {
-                return psiForUast(originalSymbol)
+                return psiForUast(originalSymbol, context)
             }
         }
     }

